crates/telio-pmtu/src/linux.rs

```diff
@@ -0,0 +1,387 @@
+use std::{ffi::c_int, io, net::IpAddr, os::fd::AsRawFd, time::Duration};
+
+use pnet_packet::{icmp, icmpv6, Packet};
+use telio_utils::{telio_log_debug, telio_log_warn};
+use tokio::net::UdpSocket;
+
+pub struct PMTUSocket {
+    sock: UdpSocket,
+    host: IpAddr,
+    response_timeout: Duration,
+}
+
+const RETRIES: u32 = 3;
+
+impl PMTUSocket {
+    pub fn new(host: IpAddr, response_timeout: Duration) -> io::Result<Self> {
+        let sock = crate::create_icmp_socket(&host)?;
+        set_pmtu_on(&sock, host.is_ipv6())?;
+        Ok(Self {
+            sock,
+            host,
+            response_timeout,
+        })
+    }
+
+    pub async fn probe_pmtu(&self) -> io::Result<u32> {
+        telio_log_debug!("Connecting to: {}", self.host);
+        self.sock.connect((self.host, 0)).await?;
+
+        telio_log_debug!("Connected to: {}", self.host);
+
+        let mut pmtu = read_pmtu(&self.sock, self.host.is_ipv6()).unwrap_or_else(|err| {
+            telio_log_warn!("Failed to fetch initial kernel PMTU estimate: {err:?}");
+            (super::PMTU_RANGE.start + super::PMTU_RANGE.end) / 2
+        });
+
+        if !super::PMTU_RANGE.contains(&pmtu) {
+            // PMTU is not contained within our reasonable range,
+            // let's probe from the top of the range
+            pmtu = super::PMTU_RANGE.end - 1;
+        }
+
+        telio_log_debug!("Initial PMTU estimate: {pmtu}");
+
+        let mut msgbuf = vec![0xFF; super::PMTU_RANGE.end as usize];
+
+        if self.host.is_ipv6() {
+            let mut pkg = IcmpV6::new(&mut msgbuf)?;
+            self.probe_pmtu_internal(&mut pkg, pmtu).await
+        } else {
+            let mut pkg = IcmpV4::new(&mut msgbuf)?;
+            self.probe_pmtu_internal(&mut pkg, pmtu).await
+        }
+    }
+
+    async fn probe_pmtu_internal(&self, pkg: &mut impl IcmpPkg, mut pmtu: u32) -> io::Result<u32> {
+        pkg.set_ident(super::echo_req_ident());
+
+        let mut pmtu_range = super::PMTU_RANGE;
+
+        loop {
+            let check = self.is_pmtu_ok(pkg, pmtu).await?;
+            telio_log_debug!("PMTU {pmtu} result: {check:?}");
+
+            pmtu = match check {
+                PmtuResult::Ok => {
+                    pmtu_range.start = pmtu;
+                    (pmtu_range.start + pmtu_range.end) / 2
+                }
+                PmtuResult::Dropped => {
+                    pmtu_range.end = pmtu;
+                    (pmtu_range.start + pmtu_range.end) / 2
+                }
+                PmtuResult::FragNeeded => {
+                    // Read the kernel PMTU estimate, since the kernel has the knowledge of error ICMP packet.
```

Comment:
:thinking: Maybe it is worth considering totreat `PmtuResult::FragNeeded` as simply `Pmtu:Result::Dropped`? 

It seems like here the only benefit we gain from using ICMP next_hop MTU field is a slightly faster search of the PMTU. But on the other hand, non-cooperating node (e.g. non-complient by bug or malicious) has ability to cause this loop to spin infinitely. Which seems like not worth in terms of only few packets saved per hour :thinking:

Reply:
I do agree that we can't fully trust the kernel. But its hints are beneficial for the speed of PMTU discovery.
So I see a different solution. Let's do the kernel estimate but break the loop after reaching some number of turns. In that way, we get the best of both worlds

Reply:
I've added the cap of the number of iterations

Reply:
:heavy_check_mark: